File: Backend/src/main/java/com/github/francisfire/anavis/controller/DonorController.java
package com.github.francisfire.anavis.controller;

import java.util.Set;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.github.francisfire.anavis.component.AccessCheckerComponent;
import com.github.francisfire.anavis.models.Donor;
import com.github.francisfire.anavis.services.DonorServices;

@RestController
@RequestMapping("api/donor")
public class DonorController {

	@Autowired
	private DonorServices donorServices;
	@SuppressWarnings("unused")
	@Autowired
	private AccessCheckerComponent accessCheckerComponent;

	@PreAuthorize("hasAuthority('OFFICE') or @accessCheckerComponent.sameDonorId(principal, #donorMail)")
	@GetMapping("/{donorMail}/canDonate")
	public boolean checkDonationPossibility(@PathVariable("donorMail") String donorMail) {
		return donorServices.checkDonationPossibility(donorMail);
	}

	@GetMapping("/office/{officeMail}/available")
	public Set<Donor> getAvailableDonorsByOfficeId(@PathVariable("officeMail") String officeMail) {
		return donorServices.getAvailableDonorsByOfficeId(officeMail);
	}
	
	@GetMapping("/{donorMail}") public Donor getDonorByMail(@PathVariable("donorMail") String donorMail) {
		return donorServices.getDonorInstance(donorMail);
	}

}

File: Frontend/lib/apicontrollers/donor_controller.dart
import 'dart:async';

import 'package:anavis/providers/app_state.dart';
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import 'package:http/http.dart' as http;

class DonorController {
  String _ip;
  String _baseUrl;
  Map<String, String> _header;

  DonorController(BuildContext context) {
    _ip = AppState().getIp();
    _header = AppState().getHttpHeaders();

    _baseUrl = "http://$_ip:8080/api/donor";
  }

  Future<String> checkDonationPossibility(String donorMail) async {
    http.Response res = await http.get(
      "$_baseUrl/$donorMail/canDonate",
      headers: this._header,
    );
    return res.body;
  }

  Future<String> getAvailableDonorsByOfficeId(String officeMail) async {
    http.Response res = await http.get(
      "$_baseUrl/office/$officeMail/available",
      headers: this._header,
    );
    return res.body;
  }

  Future<String> getDonorByMail(String donorMail) async {
    http.Response res = await http.get(
      "$_baseUrl/$donorMail",
      headers: this._header,
    );
    return res.body;
  }
}
